torch/autograd/gradcheck.py

```diff
@@ -250,58 +243,89 @@ def prepped_input(input: torch.Tensor, maybe_perturbed_input: Optional[torch.Ten
         # fn (in the gradgrad case) needs to compute grad wrt input
         return input
 
 
-def check_outputs_same_dtype_and_shape_in_neighborhood(output1, output2, idx, eps) -> None:
+def check_outputs_same_dtype_and_shape(output1, output2, eps, idx=None) -> None:
     # Check that the returned outputs don't have different dtype or shape when you
     # perturb the input
+    on_index = "on index {idx} " if idx is not None else ""
     assert output1.shape == output2.shape, \
         (f"Expected `func` to return outputs with the same shape"
-         f" when inputs are perturbed on index {idx} by {eps}, but got:"
+         f" when inputs are perturbed {on_index}by {eps}, but got:"
          f" shapes {output1.shape} and {output2.shape}.")
     assert output1.dtype == output2.dtype, \
         (f"Expected `func` to return outputs with the same dtype"
-         f" when inputs are perturbed on index {idx} by {eps}, but got:"
+         f" when inputs are perturbed {on_index}by {eps}, but got:"
          f" dtypes {output1.dtype} and {output2.dtype}.")
 
 
-def get_numerical_jacobian_wrt_specific_input(fn, input, input_idx, inputs, outputs, eps,
+def get_numerical_jacobian_wrt_specific_input(fn, input_idx, inputs, outputs, eps,
                                               grad_out) -> Tuple[torch.Tensor, ...]:
     # Computes the numerical jacobians wrt to a single input. Returns N jacobian
-    # tensors, where N is the number of outputs. Input must require grad.
-    assert input.requires_grad
-    # We need a dictionary because for sparse inputs, d_idx aren't necessarily consecutive
-    # the ith entry of jacobian_cols[j] is the jth column of jacobian w.r.t. the ith output
-    # and input at `input_idx`. The ith entry may be None for when grad_out is 1j but the ith output
-    # is not complex.
+    # tensors, where N is the number of outputs
+    # We use a dictionary because for sparse inputs, d_idx aren't necessarily consecutive
     jacobian_cols: Dict[int, List[torch.Tensor]] = {}
-
+    input = inputs[input_idx]
+    assert input.requires_grad
     for x, idx, d_idx in iter_tensor(input):
-        def wrapped_fn():
-            inp = tuple(prepped_input(a, x if i == input_idx else None) if is_tensor_like(a) else a
-                        for i, a in enumerate(_as_tuple(inputs)))
-            return tuple(a.clone() for a in _as_tuple(fn(*inp)))
-
+        wrapped_fn = with_prepped_inputs(fn, inputs, input_idx, x)
         input_to_perturb = x[idx]
-        nbhd_checks_fn = functools.partial(check_outputs_same_dtype_and_shape_in_neighborhood, idx=idx, eps=eps)
+        nbhd_checks_fn = functools.partial(check_outputs_same_dtype_and_shape, idx=idx, eps=eps)
+        jvp_fn = get_jvp_fn(wrapped_fn, input_to_perturb, eps, nbhd_checks_fn)
+        jacobian_cols[d_idx] = compute_numerical_jacobian_cols(jvp_fn, eps, x.is_complex(), grad_out)
+    return combine_jacobian_cols(jacobian_cols, outputs, input, input.numel())
```

Comment:
Did some cleanup here: moving the nested functions out allowed us to reuse between fast and slow versions and clean up the function body